jenkinsapi/jenkins.py

```diff
@@ -463,6 +463,24 @@ def _wait_for_reboot(self):
                "failure may cause other failures.")
         log.critical(msg, count * wait)
 
+    def safe_exit(self, wait_for_exit=True):
+        """ restarts jenkins when no jobs are running """
+        # NB: unlike other methods, the value of resp.status_code
+        # here can be 503 even when everything is normal
+        url = '%s/safeExit' % (self.baseurl,)
+        valid = self.requester.VALID_STATUS_CODES + [503, 500]
+        resp = self.requester.post_and_confirm_status(url, data='',
+                                                      valid=valid)
+        if wait_for_exit:
+            self._wait_for_exit()
+        return resp
+
+    def _wait_for_exit(self):
+        # We need to make sure all jobs have finished,
+        # and that jenkins is unavailable
+        # One way to be sure is to make sure jenkins is really down.
+        self.__jenkins_is_unavailable()  # Blocks until jenkins is unavailable
```

Comment:
I don't think that this code will block if Jenkins is available -  it will return immediately if Jenkins still replies with 200, 503 and 500. it is better to add actual check that will exit only on `ConnectionError`.